File: archive/co2.py
SERBIA_PERCENT_FROM_COAL = 0.7
SERBIA_PERCENT_FROM_HYDRO = 0.3
KG_OF_CO2_PER_KWH_COAL = 0.9 # kg / kwh
KG_OF_CO2_PER_KWH_HYDRO = 0.024 # kg / kwh
KG_OF_CO2_PER_KWH_SOLAR = 0.04 # kg / kwh
ADULT_TREE_ABSORBTION = 50 # kg / year
CO2_CAR_PER_100_KM = 13.4 # kg / 100 km
    
class CO2:
    """ Object that calculates annualized co2 reduction and tree/km equivalents. Provide annual values for usage and solar production. """

    
    def __init__(self, annual_usage: int, annual_solar_production:int):
        
        self.annual_co2_output_pre_solar = annual_usage * (SERBIA_PERCENT_FROM_COAL * KG_OF_CO2_PER_KWH_COAL)
        
        self.annual_co2_output_post_solar = (annual_usage - annual_solar_production) * SERBIA_PERCENT_FROM_COAL * KG_OF_CO2_PER_KWH_COAL
        
        self.reduction_kg = int(self.annual_co2_output_pre_solar - self.annual_co2_output_post_solar)
        self.number_of_trees_equivalent = int(self.reduction_kg / ADULT_TREE_ABSORBTION)
        self.car_kilometres_equivalent = int(self.reduction_kg / CO2_CAR_PER_100_KM) * 100

File: solar-proposal/lib/co2.ts
// CO2 reduction calculator
// Ported from co2.py

import {
  SERBIA_PERCENT_FROM_COAL,
  KG_OF_CO2_PER_KWH_COAL,
  ADULT_TREE_ABSORPTION_KG_YEAR,
  CO2_CAR_PER_100_KM,
} from "./constants";

export interface CO2Result {
  annualCO2OutputPreSolar: number;
  annualCO2OutputPostSolar: number;
  reductionKg: number;
  numberOfTreesEquivalent: number;
  carKilometresEquivalent: number;
}

/**
 * Calculate annual CO2 reduction from solar installation
 *
 * @param annualUsage - Annual electricity usage in kWh
 * @param annualSolarProduction - Annual solar production in kWh
 */
export function calculateCO2(
  annualUsage: number,
  annualSolarProduction: number
): CO2Result {
  // Serbia's grid emissions (mainly from coal)
  const annualCO2OutputPreSolar =
    annualUsage * SERBIA_PERCENT_FROM_COAL * KG_OF_CO2_PER_KWH_COAL;

  const annualCO2OutputPostSolar =
    (annualUsage - annualSolarProduction) *
    SERBIA_PERCENT_FROM_COAL *
    KG_OF_CO2_PER_KWH_COAL;

  const reductionKg = Math.round(annualCO2OutputPreSolar - annualCO2OutputPostSolar);

  // Equivalent to planting this many trees
  const numberOfTreesEquivalent = Math.round(reductionKg / ADULT_TREE_ABSORPTION_KG_YEAR);

  // Equivalent to this many km driven by car
  const carKilometresEquivalent = Math.round((reductionKg / CO2_CAR_PER_100_KM) * 100);

  return {
    annualCO2OutputPreSolar,
    annualCO2OutputPostSolar,
    reductionKg,
    numberOfTreesEquivalent,
    carKilometresEquivalent,
  };
}
